File: src/api/assignment-upload-file.ts
import type { FormDataFile } from "@literate.ink/utilities";
import { RequestFN } from "~/core/request-function";
import { RequestUpload } from "~/core/request-upload";
import { TabLocation, type SessionHandle } from "~/models";
import { DocumentKind } from "~/models/document-kind";
import { EntityState } from "~/models/entity-state";
import { createEntityID } from "~/api/helpers/entity-id";

export const assignmentUploadFile = async (session: SessionHandle, assignmentID: string, file: FormDataFile, fileName: string): Promise<void> => {
  // Check if the file can be uploaded.
  // Otherwise we'll get an error during the upload.
  // @ts-expect-error : trust the process.
  const fileSize: number | undefined = file.size || file.byteLength;
  const maxFileSize = session.user.authorizations.maxAssignmentFileUploadSize;
  if (typeof fileSize === "number" && fileSize > maxFileSize) {
    throw new Error(`File size is too big, maximum allowed is ${maxFileSize} bytes.`);
  }

  // Ask to the server to store the file for us.
  const fileUpload = new RequestUpload(session, "SaisieTAFARendreEleve", file, fileName);
  await fileUpload.send();

  // Now we can link the file to the assignment.
  const request = new RequestFN(session, "SaisieTAFARendreEleve", {
    _Signature_: { onglet: TabLocation.Assignments },

    donnees: {
      listeFichiers: [{
        E: EntityState.CREATION,
        G: DocumentKind.FILE,
        L: fileName,
        N: createEntityID(),
        idFichier: fileUpload.id,
        TAF: { N: assignmentID }
      }]
    }
  });

  await request.send();
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/assignment-upload-file.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.helpers.createEntityID
import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.core.RequestUpload
import ink.literate.pawnote.models.DocumentKind
import ink.literate.pawnote.models.EntityState
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import ink.literate.pawnote.models.errors.UploadSizeError
import java.io.File
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun assignmentUploadFile(
    session: SessionHandle,
    assignmentID: String,
    file: File,
    fileName: String
) {
  // Check if the file can be uploaded.
  // Otherwise we'll get an error during the upload.

  val fileSize = file.length()
  val maxFileSize = session.user.authorizations.maxAssignmentFileUploadSize

  if (fileSize > maxFileSize) throw UploadSizeError(maxFileSize)

  // Ask to the server to store the file for us.
  val fileUpload = RequestUpload(session, "SaisieTAFARendreEleve", file, fileName)
  fileUpload.send()

  val request =
      RequestFN(
          session.information,
          "SaisieTAFARendreEleve",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Assignments.code) }

                putJsonObject("donnees") {
                  putJsonArray("listeFichiers") {
                    addJsonObject {
                      put("E", EntityState.CREATION.code)
                      put("G", DocumentKind.FILE.code)
                      put("L", fileName)
                      put("N", createEntityID())
                      put("idFichier", fileUpload.id)
                      putJsonObject("TAF") { put("N", assignmentID) }
                    }
                  }
                }
              }))

  request.send()
}
